keps/sig-network/2433-topology-aware-hints/README.md

```diff
@@ -0,0 +1,578 @@
+  This approach could result in a situation where an endpoint on a Node is
+  delivered to a separate underprovisioned zone. The simplest approach would be
+  to disable this functionality altogether.
+- When this feature is transitioning between enabled and disabled states, there
+  will be a brief point in time where only some EndpointSlices have hints. That
+  could temporarily result in traffic being routed to a small subset of
+  endpoints. To avoid this, we only filter out endpoints that have a hint set
+  to a different zone. If a hint is not set for an endpoint, it will be included
+  by all instances of kube-proxy.
+
+## Design Details
+
+### Assumptions
+
+- Incoming traffic is proportional to the number of allocatable CPU cores in a
+  zone. Although this is an imperfect metric, it is the best available way of
+  predicting how much traffic will be received in a zone. If we are unable to
+  derive the number of allocatable cores in a zone we will fall back to the
+  number of nodes in that zone.
+- Service capacity is proportional to the number of endpoints in a zone. This
+  assumes that each endpoint has equivalent capacity. Although this is not
+  always true, it usually is. We can explore ways to deal with variable capacity
+  endpoints in the future.
+
+### Configuration
+
+A new `topology.kubernetes.io/topology-aware-routing` annotation will be
+supported on Services. This will have 2 potential values:
+
+- `Auto`: When there are a sufficient number of endpoints for the Service, the
+  EndpointSlice controller will add topology hints for each endpoint that will
+  ensure a proportional amounts are available to each zone in a cluster.
+- `Disabled`: EndpointSlice hints will not be set for this feature.
+
+When this annotation is either unspecified or not set to `Auto`, the `Disabled`
+behavior will be default. A future KEP will explore changing the default
+behavior from `Disabled` to `Auto` with a new feature gate.
+
+#### Interoperability
+
+If both this annotation and `topologyKeys` are set on a Service, `topologyKeys`
+will be given precedence and this annotation will be ignored.
+
+#### Feature Gate
+
+This functionality will be guarded by the `TopologyAwareHints` feature
+gate. This gate should not be enabled if the `ServiceTopology` gate is enabled.
+
+### API
+
+A new `TopologyHints` struct would be added to the `EndpointSlice.Endpoint`
+struct:
+
+```go
+type Endpoint struct {
+  ...
+  // hints contains information associated with how an endpoint should be
+  // consumed.
+  // +optional
+  Hints EndpointHints `json:"hints,omitempty" protobuf:"bytes,7,opt,name=hints"`
+}
+```
+
+```go
+// EndpointHints provides hints describing how an endpoint should be consumed.
+type EndpointHints struct {
+  // zone indicates the zone this endpoint should be consumed by to enable
+  // topology aware routing.
+  // +optional
+  Zone string `json:"ready,omitempty" protobuf:"bytes,1,name=ready"`
+}
+```
+
+#### Future API Expansion
```

Comment:
So your future safety here is "add a new field and treat them as mutually exclusive" ?  It seems to me that we'll need multiple zones (and probably weight) pretty quickly in order to handle small services.

e.g. 2 replicas (in zones A and B) and 3 node zones (A, B, C).  You'll need to set up zone A to send to replica A 100%, Zone B to replica B 100% and zone C to replicas A and B at 50%.  That's a simple case where weight doesn't matter but you simply can't represent an endpoint in 2 zones with the above proposal.

e.g. 4 replicas (2 in zone A, 1 each in B and C) and 3 node-zones.  You'll want something like:

zone A: 50% to replica a1; 50% to replica a2
zone B: 75% to replica b1; 25% to replica a1
zone C: 75% to replica c1; 25% to replica a2

I think it will be important quickly - am I wrong?  To that end, I'd suggest we start with `zones []struct` now and as the impl builds up consider weight.

Note this is different than general per-endpoint weight (which we should still consider on its own), because this is a hint to consumers how to split for capacity (based on the set of endpoints), not an indicator of capacity (which is more intrinsic to the endpoint).  I'm not sure that holds water yet, ergo let's not add weight yet :)

Reply:
The answer for the current proposal is that we simply would fallback to slices without hints if we could not reach a reasonable balance otherwise. When we tested weighting with the previous proposals it looked like it would significantly increase the number of EndpointSlice updates. Weight is most useful when there are a small number of endpoints. This is also when topology aware routing includes the most risk since it would be very easy for an individual endpoint to be overloaded. So even if we had weight and could accomplish topology aware routing with only 2 endpoints, the result might not be great. Agree that it's worth expanding the API to make weight easy to add in the future and also agree that we shouldn't add weight yet.

Reply:
MY argument is to start now with `zones []struct` with no weight and when (inevitably, IMO) we get complaints that small-N services don't work, we can revisit weight.  it might be, as you say, that weight only matters at the very low end (N < 12), but that's a lot of Services.  All your harm-reduction heuristics will have to consider it, so it's definitely a different discussion.